### DISCUSSION TITLE
Designing dependencies in multi-layered architecture

### DISCUSSION THREAD

**Discussion by tomassliz** [pointfreeco/episode-code-samples#74] [Q&A] [answered]
Hello,
I just saw the "Designing dependencies" series and I can't figure out how to adopt principles from it in our multi-layered architecture (commonly known as "clean architecture"). You are setting dependencies in view model and view model is observable. So you have basically 3 layers:
- presentation (SwiftUI)
- view models (observables)
- dependencies

But consider please the example bellow where I have one more layer - repositories:

```swift
protocol CarsRepositoryProtocol {
    func fetchCars() -> AnyPublisher<[Car], Error>
}

class CarsRepository: CarsRepositoryProtocol {
    private let carsService: CarsRemoteServiceProtocol
    private let databaseService: DatabaseServiceProtocol

    init(
        carsService: CarsRemoteServiceProtocol,
        databaseService: DatabaseServiceProtocol
    ) {
        self.carsService = carsService
        self.databaseService = databaseService
    }

    func fetchCars() -> AnyPublisher<[Car], Error> {
        ...
    }
}

class GarageViewModel: GarageViewModelProtocol {
    private let carsRepository: CarsRepositoryProtocol

    init(carsRepository: CarsRepositoryProtocol) {
        self.carsRepository = carsRepository
    }
}
```

How can I set dependencies for the repository and implement the `fetchCars()` at the same time when I want to use protocol witness instead protocol? I can have something like this, but it feels like cumbersome:

```swift
struct CarsRepository {
    let carsService: CarsRemoteService
    let databaseService: DatabaseService

    var cars: () -> AnyPublisher<[Car], Error>

    init(
        carsService: CarsRemoteService,
        databaseService: DatabaseService,
        cars: @escaping () -> AnyPublisher<[Car], Error>
    ) {
        self.carsService = carsService
        self.databaseService = databaseService
        self.cars = cars
    }
}

// Mocks
extension CarsRepository {
    static func empty(
        carsService: CarsRemoteService,
        databaseService: DatabaseService
    ) -> Self {
        Self(
            carsService: carsService,
            databaseService: databaseService,
            cars: {
                Just([])
                    .setFailureType(to: Error.self)
                    .eraseToAnyPublisher()
            }
        }
    }

    static func happyPath(
        carsService: CarsRemoteService,
        databaseService: DatabaseService
    ) -> Self {
        ...
    }
}
```

Maybe I'm totally wrong and I totally missed your message. I'm curious anyway. Thanks for your time 🙏

**Comment by mbrandonw** [comment, marked as answer]
It's not clear what `CarsRemoteService` and `DatabaseService` are. Right now they are just being passed around but never actually used. What is their functionality? Can you provide a bit more code on how they are defined and used in `CarsRepository`?

**Reply by tomassliz** [reply]
I don't have code for them because this is only simple example. You can think about them in this way:
- `CarsRemoteService` is something like `WeatherClient` in your example. It communicates with some endpoint and retrieves data.
- `DatabaseService` saves retrieved data in some kind of database or on disk, it doesn't matter.

Both of these protocols can be replaced with simple protocol witnesses.

**Reply by mbrandonw** [reply]
It's hard for me to show how this can be done in the style we describe in the episodes without knowing how those objects are supposed to behave. If I ignore those objects, then to me the real dependency is just this:

```swift
struct CarsClient {
  var fetch: () -> AnyPublisher<[Car], Error>
}
```

Then a "no-op" client is this:

```swift
extension CarsClient {
  static let noop = Self(fetch: { Empty().eraseToAnyPublisher() })
}
```

And a "live" client would do the real work to actually fetch cars. If that work is done via a network request, it could be:

```swift
extension CarsClient {
  static let live = Self(
    fetch: { 
      URLSession.shared.dataTaskPublisher(...)
        .map { data, _ in data }
        .decode([Car].self)
        .eraseToAnyPublisher()
    }
  )
}
```

If the work is done comes from a database dependency, then it could be:

```swift
extension CarsClient {
  static func live(database: Database) -> Self {
    Self(
      fetch: {
        database.fetchCars()
      }
    )
  }
}
```

If what you are talking about is more like the latter, then I would ask why is the `CarsClient` needed at all? Why not have a `DatabaseClient` with a `fetchCars` endpoint that does the work?

So again, it's hard for me to provide a targeted answer without knowing how you want to use `CarsRemoteService` and `DatabaseService` because in my opinion you probably don't need three separate things: `CarsRemoteService`, `DatabaseService` _and_ `CarRepository`.

**Reply by tomassliz** [reply]
I understand and I appreciate your answer. My example is not very clever but I get the idea how to pass `Database` dependency into `CarsClient`.
It would be great to see this concept in real world app. My situation is probably different because I have to live in more complicated environment which is presented with multiple layers and each of these layers depends on previous one. So I have pass these dependencies through these layers ...
You, on the other hand, have very flat structure so it's easy to use dependencies as you showed.
